// react-native-localize/dist/commonjs/extras/expo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withLocalize = exports.default = void 0;
var _configPlugins = require("@expo/config-plugins");
var _generateCode = require("@expo/config-plugins/build/utils/generateCode");
var _fs = require("fs");
var _path = require("path");
const PACKAGE_NAME = "react-native-localize";
const withIosAppLocales = ({
  ios = {},
  ...config
}, locales) => ({
  ...config,
  ios: {
    ...ios,
    infoPlist: {
      ...ios.infoPlist,
      CFBundleLocalizations: locales
    }
  }
});
const withAndroidAppLocales = (config, locales) => {
  const withLocalesConfig = (0, _configPlugins.withDangerousMod)(config, ["android", config => {
    const xmlDir = (0, _path.join)(config.modRequest.platformProjectRoot, "app", "src", "main", "res", "xml");
    (0, _fs.mkdirSync)(xmlDir, {
      recursive: true
    });
    (0, _fs.writeFileSync)((0, _path.join)(xmlDir, "locales_config.xml"), `<?xml version="1.0" encoding="utf-8"?>
<locale-config xmlns:android="http://schemas.android.com/apk/res/android">
${locales.map(locale => `  <locale android:name="${locale}"/>`).join("\n")}
</locale-config>
`);
    return config;
  }]);
  const withMainApplication = (0, _configPlugins.withAndroidManifest)(withLocalesConfig, config => {
    const mainApplication = _configPlugins.AndroidConfig.Manifest.getMainApplicationOrThrow(config.modResults);
    mainApplication.$ = {
      ...mainApplication.$,
      "android:localeConfig": "@xml/locales_config"
    };
    return config;
  });
  return (0, _configPlugins.withAppBuildGradle)(withMainApplication, config => {
    const {
      modResults
    } = config;
    const list = locales.reduce((acc, item) => {
      const [language, scriptOrRegion, region] = item.split("-");
      if (language == null) {
        return acc;
      }

      // https://developer.android.com/guide/topics/resources/app-languages#gradle-config
      if (scriptOrRegion != null && region != null) {
        return [...acc, `b+${language}+${scriptOrRegion}+${region}`];
      }
      if (scriptOrRegion != null) {
        return [...acc, `${language}-r${scriptOrRegion}`];
      }
      return [...acc, language];
    }, []).filter((item, index, array) => array.indexOf(item) === index).map(item => `"${item}"`).join(", ");
    const {
      contents
    } = (0, _generateCode.mergeContents)({
      src: modResults.contents,
      comment: "//",
      tag: PACKAGE_NAME,
      offset: 1,
      anchor: /versionName "[\d.]+"/,
      newSrc: `        resourceConfigurations += ${modResults.language === "kt" ? `listOf(${list})` : `[${list}]`}`
    });
    return {
      ...config,
      modResults: {
        ...modResults,
        contents
      }
    };
  });
};
const plugin = (config, {
  locales = []
} = {}) => {
  const plugins = [];
  const {
    platforms = []
  } = config;
  const {
    android,
    ios
  } = Array.isArray(locales) ? {
    android: locales,
    ios: locales
  } : {
    android: locales.android ?? [],
    ios: locales.ios ?? []
  };
  if (platforms.includes("ios") && ios.length > 0) {
    plugins.push([withIosAppLocales, ios]);
  }
  if (platforms.includes("android") && android.length > 0) {
    plugins.push([withAndroidAppLocales, android]);
  }
  return (0, _configPlugins.withPlugins)(config, plugins);
};
const withLocalize = exports.withLocalize = (0, _configPlugins.createRunOncePlugin)(plugin, PACKAGE_NAME);
var _default = config => [PACKAGE_NAME, config];
exports.default = _default;
