// react-native-localize/dist/commonjs/module.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usesMetricSystem = exports.usesAutoTimeZone = exports.usesAutoDateAndTime = exports.uses24HourClock = exports.useLocalize = exports.openAppLanguageSettings = exports.getTimeZone = exports.getTemperatureUnit = exports.getNumberFormatSettings = exports.getLocales = exports.getCurrencies = exports.getCountry = exports.getCalendar = exports.findBestLanguageTag = exports.ServerLanguagesProvider = void 0;
var _NativeRNLocalize = _interopRequireDefault(require("./specs/NativeRNLocalize.js"));
var _utils = require("./utils.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const getCalendar = () => _NativeRNLocalize.default.getCalendar();
exports.getCalendar = getCalendar;
const getCountry = () => _NativeRNLocalize.default.getCountry();
exports.getCountry = getCountry;
const getCurrencies = () => _NativeRNLocalize.default.getCurrencies();
exports.getCurrencies = getCurrencies;
const getLocales = () => _NativeRNLocalize.default.getLocales();
exports.getLocales = getLocales;
const getNumberFormatSettings = () => _NativeRNLocalize.default.getNumberFormatSettings();
exports.getNumberFormatSettings = getNumberFormatSettings;
const getTemperatureUnit = () => _NativeRNLocalize.default.getTemperatureUnit();
exports.getTemperatureUnit = getTemperatureUnit;
const getTimeZone = () => _NativeRNLocalize.default.getTimeZone();
exports.getTimeZone = getTimeZone;
const uses24HourClock = () => _NativeRNLocalize.default.uses24HourClock();
exports.uses24HourClock = uses24HourClock;
const usesMetricSystem = () => _NativeRNLocalize.default.usesMetricSystem();
exports.usesMetricSystem = usesMetricSystem;
const usesAutoDateAndTime = () => _NativeRNLocalize.default.usesAutoDateAndTime() ?? undefined;
exports.usesAutoDateAndTime = usesAutoDateAndTime;
const usesAutoTimeZone = () => _NativeRNLocalize.default.usesAutoTimeZone() ?? undefined;
exports.usesAutoTimeZone = usesAutoTimeZone;
const findBestLanguageTag = exports.findBestLanguageTag = (0, _utils.getFindBestLanguageTag)(getLocales());
const openAppLanguageSettings = async () => {
  await _NativeRNLocalize.default.openAppLanguageSettings();
};
exports.openAppLanguageSettings = openAppLanguageSettings;
const ServerLanguagesProvider = ({
  children
}) => children;
exports.ServerLanguagesProvider = ServerLanguagesProvider;
const api = {
  getCalendar,
  getCountry,
  getCurrencies,
  getLocales,
  getNumberFormatSettings,
  getTemperatureUnit,
  getTimeZone,
  uses24HourClock,
  usesMetricSystem,
  usesAutoDateAndTime,
  usesAutoTimeZone,
  findBestLanguageTag,
  openAppLanguageSettings
};
const useLocalize = () => api;
exports.useLocalize = useLocalize;
