// react-native-localize/dist/commonjs/extras/mock/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usesMetricSystem = exports.usesAutoTimeZone = exports.usesAutoDateAndTime = exports.uses24HourClock = exports.useLocalize = exports.openAppLanguageSettings = exports.getTimeZone = exports.getTemperatureUnit = exports.getNumberFormatSettings = exports.getLocales = exports.getCurrencies = exports.getCountry = exports.getCalendar = exports.findBestLanguageTag = exports.ServerLanguagesProvider = void 0;
const getCalendar = () => "gregorian"; // the calendar identifier you want
exports.getCalendar = getCalendar;
const getCountry = () => "US"; // the country code you want
exports.getCountry = getCountry;
const getCurrencies = () => ["USD", "EUR"]; // can be empty array
exports.getCurrencies = getCurrencies;
const getLocales = () => [
// extend if needed, add the locales you want
{
  countryCode: "US",
  languageTag: "en-US",
  languageCode: "en",
  isRTL: false
}, {
  countryCode: "FR",
  languageTag: "fr-FR",
  languageCode: "fr",
  isRTL: false
}];
exports.getLocales = getLocales;
const getNumberFormatSettings = () => ({
  decimalSeparator: ".",
  groupingSeparator: ","
});
exports.getNumberFormatSettings = getNumberFormatSettings;
const getTemperatureUnit = () => "celsius"; // or "fahrenheit"
exports.getTemperatureUnit = getTemperatureUnit;
const getTimeZone = () => "Europe/Paris"; // the timezone you want
exports.getTimeZone = getTimeZone;
const uses24HourClock = () => true;
exports.uses24HourClock = uses24HourClock;
const usesMetricSystem = () => true;
exports.usesMetricSystem = usesMetricSystem;
const usesAutoDateAndTime = () => true;
exports.usesAutoDateAndTime = usesAutoDateAndTime;
const usesAutoTimeZone = () => true;

// use a provided translation, or return undefined to test your fallback
exports.usesAutoTimeZone = usesAutoTimeZone;
const findBestLanguageTag = () => ({
  languageTag: "en-US",
  isRTL: false
});
exports.findBestLanguageTag = findBestLanguageTag;
const openAppLanguageSettings = async () => {};
exports.openAppLanguageSettings = openAppLanguageSettings;
const ServerLanguagesProvider = ({
  children
}) => children;
exports.ServerLanguagesProvider = ServerLanguagesProvider;
const useLocalize = () => ({
  getCalendar,
  getCountry,
  getCurrencies,
  getLocales,
  getNumberFormatSettings,
  getTemperatureUnit,
  getTimeZone,
  uses24HourClock,
  usesMetricSystem,
  usesAutoDateAndTime,
  usesAutoTimeZone,
  findBestLanguageTag,
  openAppLanguageSettings
});
exports.useLocalize = useLocalize;
